File: 0003-longest-substring-without-repeating-characters/0003-longest-substring-without-repeating-characters.cpp
class Solution {
public:
    int lengthOfLongestSubstring(string s) {
        int n = s.size();
        if (n == 0) return 0;
        
        int start = 0, end = 0, ans = 0;
        map<char, int> mp;
        
        while (end < n) {
            mp[s[end]]++;  
            
          
            while (mp[s[end]] > 1) {
                mp[s[start]]--;
                if (mp[s[start]] == 0) {
                    mp.erase(s[start]);
                }
                start++;
            }
            
            ans = max(ans, end - start + 1);
            end++;
        }
        
        return ans;
    }
};

File: 0003-longest-substring-without-repeating-characters/0003-longest-substring-without-repeating-characters.py
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        mp = {}
        n = len(s)
        if n == 0:
            return 0
        start=0
        end = 0
        ans = 0
        while end < n:
            if s[end] in mp:
                mp[s[end]] = mp[s[end]] + 1;
            else :
                mp[s[end]] = 1
            while mp[s[end]]>1:
                mp[s[start]]-=1;
                if mp[s[start]] == 0:
                    del mp[s[start]]
                
                start+=1
            
            ans = max(ans,end-start+1)
            end+=1
        return ans
            